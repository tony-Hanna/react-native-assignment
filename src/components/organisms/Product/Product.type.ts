export type productProp = {
    _id: string,
    title:string,
    description:string,
    price:number,
    images: {  
        url: string,
        _id: string
    }[]
}