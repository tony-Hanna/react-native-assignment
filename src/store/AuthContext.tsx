import React, { createContext, useContext, useState } from 'react';

const AuthContext = createContext<{
  isAuthenticated: boolean;
  setIsAuthenticated: (auth: boolean) => void;
}>({
  isAuthenticated: false,
  setIsAuthenticated: () => {},
});

export const AuthProvider = ({ children }: { children: React.ReactNode }) => {
  const [isAuthenticated, setIsAuthenticated] = useState(false);

  return (
    <AuthContext.Provider value={{ isAuthenticated, setIsAuthenticated }}>
      {children}
    </AuthContext.Provider>
  );
};

export const useAuth = () => useContext(AuthContext);
