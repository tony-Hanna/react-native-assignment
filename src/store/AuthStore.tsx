import { create } from 'zustand';
import RNSecureStorage, {ACCESSIBLE} from 'rn-secure-storage';
import { createJSONStorage, persist, StateStorage } from 'zustand/middleware';

const SecureStorage: StateStorage = {
  setItem: async (name, value) => {
    try {
      await RNSecureStorage.setItem(name, value, {
        accessible: ACCESSIBLE.WHEN_UNLOCKED,
      });
    } catch (error) {
      console.error('Error setting secure storage item:', error);
    }
  },

  getItem: async name => {
    try {
      const value = await RNSecureStorage.getItem(name);
      return value ?? null;
    } catch (error: any) {
      if (error?.message?.includes(`Value for ${name} does not exist`)) {
         console.log(`Secure storage item '${name}' not found on startup.`);
         return null;
      } 
      console.error('Error getting secure storage item:', error);
      return null;
    }
  },

  removeItem: async name => {
    try {
      await RNSecureStorage.removeItem(name);
    } catch (error) {
      console.error('Error removing secure storage item:', error);
    }
  },
};

type AuthState = {
  userId: string | null;
  accessToken: string | null;
  refreshToken: string | null;
  setTokens: (accessToken: string, refreshToken: string) => void;
  clearTokens: () => void;
  hasStoreLoaded: boolean;
  setHasStoreLoaded: () => void;
  setUserId: (userId: string) => void;
}

export const useAuthStore = create<AuthState>()(
  persist(
    (set) => ({
      accessToken: null,
      refreshToken: null,
      userId: null,
      hasStoreLoaded: false,
      setHasStoreLoaded: () => set({ hasStoreLoaded: true }),
      setTokens: (accessToken, refreshToken) =>
        set({ accessToken, refreshToken }),
      clearTokens: () => set({ accessToken: null, refreshToken: null }),
      setUserId: (userId: string) => set({ userId }),
    }),
    {
      name: 'auth-storage',
      onRehydrateStorage: () => (state) => {
        // Ensure we always set hasStoreLoaded to true, even if rehydration fails
        if (state) {
          state.setHasStoreLoaded();
        } else {
          // If state is null, something went wrong with rehydration
          console.warn('Auth store rehydration failed');
          useAuthStore.getState().setHasStoreLoaded();
        }
      },
      storage: createJSONStorage(() => SecureStorage),
    },
  ),
);
