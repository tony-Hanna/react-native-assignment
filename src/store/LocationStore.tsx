import { create } from 'zustand';
import { Region } from 'react-native-maps';

type LocationStore = {
  location: Region | null;
  setLocation: (location: Region) => void;
  clearLocation: () => void;
};

export const useLocationStore = create<LocationStore>((set) => ({
  location: null,
  setLocation: (location) => set({ location }),
  clearLocation: () => set({ location: null }),
}));
