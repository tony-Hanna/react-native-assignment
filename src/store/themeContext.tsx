import { useState, useContext, createContext, useMemo, useCallback, memo } from "react";

export interface Theme {
  background: string;
  text: string;
  gradient: string[]
}

interface ThemeContextType {
  theme: Theme;
  isDark: boolean;
  toggleTheme: () => void;
}

const ThemeContext = createContext<ThemeContextType>({} as ThemeContextType);

const lightTheme: Theme = {
  background: "#ffffff",
  text: "#000000",
  gradient : ['#fde6d5', '#dfd4ff']
};

const darkTheme: Theme = {
  background: "#000000",
  text: "#ffffff",
  gradient : ['#0f2027', '#243B55']
};

export const ThemeProvider = ({ children }: { children: React.ReactNode }) => {
  const [isDark, setIsDark] = useState(false);

  const toggleTheme = () => setIsDark(prev => !prev);


  const theme = isDark ? darkTheme : lightTheme;

  const contextValue = useMemo(() => ({
    theme,
    isDark,
    toggleTheme
  }), [theme, isDark, toggleTheme]);

  return (
    <ThemeContext.Provider value={contextValue}>
      {children}
    </ThemeContext.Provider>
  );
};

export const useTheme = () => useContext(ThemeContext);
  