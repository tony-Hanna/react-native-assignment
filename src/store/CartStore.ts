import { create } from "zustand";
import { MMKV } from 'react-native-mmkv';
import { persist, createJSONStorage } from 'zustand/middleware';
import { CartItem } from "../screens/Cart/Cart.type";
import { useAuthStore } from "./AuthStore";
// Initialize MMKV storage
const storage = new MMKV();

// Helper to get the current user-specific storage key
const getCartStorageKey = () => {
  // Use getState to avoid hook-in-hook issues
  const userId = require('./AuthStore').useAuthStore.getState().userId;
  return userId ? `cart-storage-${userId}` : 'cart-storage-guest';
};

// Define the cart store state and actions
interface CartStore {
  items: CartItem[];
  addItem: (item: Omit<CartItem, 'quantity'>) => void;
  removeItem: (itemId: string) => void;
  updateQuantity: (itemId: string, quantity: number) => void;
  clearCart: () => void;
  getTotalItems: () => number;
  getTotalPrice: () => number;
  isInCart: (itemId: string) => boolean;
}

// Create the cart store with user-specific persistence
export const useCartStore = create<CartStore>()(
  persist(
    (set, get) => ({
      items: [],

      addItem: (item) => {
        set((state) => {
          const existingItem = state.items.find((i) => i._id === item._id);
          
          if (existingItem) {
            // If item exists, increase quantity
            return {
              items: state.items.map((i) =>
                i._id === item._id
                  ? { ...i, quantity: i.quantity + 1 }
                  : i
              ),
            };
          }
          
          // If item doesn't exist, add new item with quantity 1
          return {
            items: [...state.items, { ...item, quantity: 1 }],
          };
        });
      },

      removeItem: (itemId) => {
        set((state) => ({
          items: state.items.filter((item) => item._id !== itemId),
        }));
      },

      updateQuantity: (itemId, quantity) => {
        if (quantity < 1) {
          get().removeItem(itemId);
          return;
        }

        set((state) => ({
          items: state.items.map((item) =>
            item._id === itemId ? { ...item, quantity } : item
          ),
        }));
      },

      clearCart: () => {
        set({ items: [] });
      },

      getTotalItems: () => {
        return get().items.reduce((total, item) => total + item.quantity, 0);
      },

      getTotalPrice: () => {
        return get().items.reduce(
          (total, item) => total + item.price * item.quantity,
          0
        );
      },

      isInCart: (itemId) => {
        return get().items.some((item) => item._id === itemId);
      },
    }),
    {
      name: getCartStorageKey(),
      storage: createJSONStorage(() => ({
        getItem: (name) => {
          const value = storage.getString(name);
          return value ? JSON.parse(value) : null;
        },
        setItem: (name, value) => {
          storage.set(name, JSON.stringify(value));
        },
        removeItem: (name) => {
          storage.delete(name);
        },
      })),
    }
  )
);