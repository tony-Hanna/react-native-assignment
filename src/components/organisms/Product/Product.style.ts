import { StyleSheet } from "react-native"
import { Theme } from "../../../store/themeContext"
import { normalizeWidth, normalizeHeight } from "../../../utils/scale"
export const createStyles = (theme: Theme, isDark: boolean) =>
    StyleSheet.create({
        imagePriceWrap: {
          display: 'flex',  
          flexDirection: 'row',
          justifyContent: 'space-between',
          alignItems: 'center'
        },
        card: {
            backgroundColor: isDark?'rgba(255, 255, 255, 0.1)': 'rgba(255, 255, 255, 0.3)',
            padding: 15,
            margin: 15,
            shadowColor: "#000",
            shadowOffset: { width: 0, height: 2 },
            shadowOpacity: 0.1,
            shadowRadius: 4,
            boxShadow: isDark?'rgba(255, 255, 255, 0.66) 0px 3px 8px':'rgba(0, 0, 0, 0.24) 0px 3px 8px'
        },
        image: {
            width: normalizeWidth(150),
            height: normalizeHeight(150),
          },
        title: {
            fontSize: 20,
            fontWeight: '600',
            width: '50%',
            color: theme.text,
            marginBottom: 4,

        },
        details: {
            fontSize: 13,
            color: isDark?theme.text:'#777',
            marginBottom: 6,
            fontFamily: 'Roboto-Regular'
        },
        price: {
            fontSize: 18,
            color: '#2e8b57',
            fontWeight: 'bold',
        },
    })
