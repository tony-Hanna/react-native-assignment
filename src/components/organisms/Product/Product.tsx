import { Image, View, Pressable } from "react-native"
import { productProp } from "./Product.type"
import { createStyles } from "./Product.style"
import { useTheme } from "../../../store/themeContext"
import { useNavigation } from "@react-navigation/native"
import { NativeStackNavigationProp } from '@react-navigation/native-stack'
import { primary } from "../../../globalStyles/global.style"
import { CustomText } from "../../atoms/CustomText/CustomText"
import { MainStackParamList } from "../../../navigation/stacks/types"
import { memo, useMemo } from "react"
import Config from "react-native-config"

const Product = memo(({
    title,
    description,
    price,
    createdAt,
    images,
    _id
}: Pick<productProp, 'title' | 'description' | 'price' | 'createdAt' | 'images' | '_id'>) => {
    const { theme, isDark } = useTheme()
    
    // Memoize styles to prevent recalculation on every render
    const styles = useMemo(() => createStyles(theme, isDark), [theme, isDark])
    
    const navigation = useNavigation<NativeStackNavigationProp<MainStackParamList>>()

    // Memoize the formatted date
    const formattedDate = useMemo(() => {
        const date = new Date(createdAt);
        return date.toLocaleDateString('en-US', { 
            month: 'short', 
            day: 'numeric',
            year: 'numeric'
        });
    }, [createdAt]);

    // Memoize the image URI to prevent string concatenation on every render
    const imageUri = useMemo(() => 
        `${Config.API_URL}${images[0]?.url}`, 
        [images]
    );

  
    const handlePress = () => {
        navigation.navigate('Details', { id: _id })
    }

    return (
        <Pressable style={[styles.card, primary.borderPrimary]} onPress={handlePress}>
            <View style={styles.imagePriceWrap}>
                <CustomText numberOfLines={2} ellipsizeMode="tail" style={styles.title}>
                    {title}
                </CustomText>
                <Image
                    source={{ uri: imageUri }}
                    style={[styles.image, primary.borderPrimary]}
                    resizeMode="contain"
                />
            </View>
            <View style={styles.bottomRow}>
                <View>
                    <CustomText style={styles.price}>${price}</CustomText>
                    <CustomText numberOfLines={1} ellipsizeMode="tail" style={styles.details}>
                        {description}
                    </CustomText>
                </View>
                <CustomText style={styles.dateText}>
                    {formattedDate}
                </CustomText>
            </View>
        </Pressable>
    )
})

export { Product }
