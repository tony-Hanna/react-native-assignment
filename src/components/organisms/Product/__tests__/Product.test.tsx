import React from 'react'
import { render, fireEvent } from '@testing-library/react-native'
import { Product } from '../Product'
import { ThemeProvider } from '../../../../store/themeContext'

// Mock navigation
const mockNavigate = jest.fn()
jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    navigate: mockNavigate
  })
}))

// Mock react-native-config
jest.mock('react-native-config', () => ({
  default: {
    API_URL: 'https://mock-api.com/'
  }
}))

const wrapper = ({ children }: any) => (
  <ThemeProvider>{children}</ThemeProvider>
)

describe('Product Component', () => {
  const props = {
    _id: '1',
    title: 'Test title',
    description: 'sample description',
    price: 25.5,
    createdAt: '2023-01-01T00:00:00Z',
    images: [{ url: '/images/sample.jpg' }]
  }

  it('matches snapshot', () => {
    const { toJSON } = render(<Product {...props} />, { wrapper });
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders product component', () => {
    const {getByText} = render(<Product {...props}/>, {wrapper})

    expect(getByText('Test title')).toBeTruthy()
    expect(getByText('sample description')).toBeTruthy()
    expect(getByText('$25.5')).toBeTruthy()
    expect(getByText('Jan 1, 2023')).toBeTruthy()
  })

  it('navigates to Details screen when pressed', () => {
    const { getByText } = render(<Product {...props} />, { wrapper })
       //  Navigate up the component tree to find the Pressable component
    // parent?.parent means:
    // - first parent: the View containing the title
    // - second parent: the Pressable component that wraps everything
    const titleElement = getByText('Test title')
    const productCard = titleElement.parent?.parent
    if (!productCard) throw new Error('Could not find product card element')
    
    fireEvent.press(productCard)

    expect(mockNavigate).toHaveBeenCalledWith('Details', { id: '1' })
  })
})
