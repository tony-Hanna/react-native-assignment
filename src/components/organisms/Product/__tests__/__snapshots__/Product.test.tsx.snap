// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Component matches snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "backgroundColor": "rgba(255, 255, 255, 0.3)",
        "boxShadow": "rgba(0, 0, 0, 0.24) 0px 3px 8px",
        "margin": 15,
        "padding": 15,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 4,
      },
      {
        "borderRadius": 10,
      },
    ]
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      ellipsizeMode="tail"
      numberOfLines={2}
      style={
        [
          {
            "color": "#000000",
            "fontFamily": "Roboto-Medium",
          },
          {
            "color": "#000000",
            "fontSize": 20,
            "fontWeight": "600",
            "marginBottom": 4,
            "width": "50%",
          },
        ]
      }
      testID="custom-text"
    >
      Test title
    </Text>
    <Image
      resizeMode="contain"
      source={
        {
          "uri": "undefined/images/sample.jpg",
        }
      }
      style={
        [
          {
            "height": 247,
            "width": 300,
          },
          {
            "borderRadius": 10,
          },
        ]
      }
    />
  </View>
  <View
    style={
      {
        "alignItems": "flex-end",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginTop": 8,
      }
    }
  >
    <View>
      <Text
        style={
          [
            {
              "color": "#000000",
              "fontFamily": "Roboto-Medium",
            },
            {
              "color": "#2e8b57",
              "fontSize": 18,
              "fontWeight": "bold",
            },
          ]
        }
        testID="custom-text"
      >
        $
        25.5
      </Text>
      <Text
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          [
            {
              "color": "#000000",
              "fontFamily": "Roboto-Medium",
            },
            {
              "color": "#777",
              "fontFamily": "Roboto-Regular",
              "fontSize": 13,
              "marginBottom": 6,
            },
          ]
        }
        testID="custom-text"
      >
        sample description
      </Text>
    </View>
    <Text
      style={
        [
          {
            "color": "#000000",
            "fontFamily": "Roboto-Medium",
          },
          {
            "color": "#666",
            "fontSize": 12,
            "fontStyle": "italic",
          },
        ]
      }
      testID="custom-text"
    >
      Jan 1, 2023
    </Text>
  </View>
</View>
`;
